src/yarp.c

```diff
@@ -2154,6 +2157,49 @@ parse_arguments_list(yp_parser_t *parser, yp_arguments_t *arguments) {
       break;
     }
   }
+
+  if (accept(parser, YP_TOKEN_BRACE_LEFT) || accept(parser, YP_TOKEN_KEYWORD_DO)) {
+    not_provided(parser);
+    not_provided(parser);
+
+    yp_token_t opening = parser->previous;
+
+    accept(parser, YP_TOKEN_NEWLINE);
+
+    yp_node_t *block_arguments = yp_node_arguments_node_create(parser);
+    if (accept(parser, YP_TOKEN_PIPE)) {
+      while (!accept(parser, YP_TOKEN_PIPE)) {
+        if (block_arguments->as.arguments_node.arguments.size != 0) {
+          expect(parser, YP_TOKEN_COMMA, "Expected a ',' in block argument list.");
+        }
+
+        yp_node_t *expression = parse_expression(parser, BINDING_POWER_BITWISE_AND, "expected an argument passed to the block.");
+        yp_node_list_append(parser, block_arguments, &block_arguments->as.arguments_node.arguments, expression);
+      }
+    }
+
+    accept(parser, YP_TOKEN_NEWLINE);
+
+    arguments->arguments =  yp_node_arguments_node_create(parser);
+
+    yp_node_t *statements;
+
+    if (accept(parser, YP_TOKEN_BRACE_RIGHT) || accept(parser, YP_TOKEN_KEYWORD_END)) {
+      statements = yp_node_statements_create(parser);
```

Comment:
Instead of creating this here, how about we make this optional?

Reply:
Additionally, we'll have to do something different for braces versus do/end because with do/end you can use `rescue`/`else`/`ensure`.